Introducing new API for Advanced Threat Protection

tomerlmsft opened the pull request:
API form -
{resourceUri}/providers/Microsoft.Security/advancedThreatProtectionSettings/{settingName}

Suppporting verbs -
GET, PUT and Delete.

URI example -
/subscriptions/20ff7fc3-e762-44dd-bd96-b71116dcdc23/resourceGroups/SampleRG/providers/Microsoft.Storage/storageAccounts/samplestorageaccount/providers/Microsoft.Security/advancedThreatProtectionSettings/current

Design doc for ATP -
https://microsoft.sharepoint.com/teams/Data_Security/Documents/Dev%20Specs/Storage%20Threat%20Detection/Threat%20Detection%20for%20Azure%20Storage%20-%20Design%20Document.docx?d=w996c55b5a160468c9a9bd0434418a4d0

VSTS item - 
#247253, #257790

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

tomerlmsft committed:
Introducing new API for Advanced Threat Protection

API form -
{resourceUri}/providers/Microsoft.Security/advancedThreatProtectionSettings/{settingName}

Suppporting verbs -
GET, PUT and Delete.

URI example -
/subscriptions/20ff7fc3-e762-44dd-bd96-b71116dcdc23/resourceGroups/SampleRG/providers/Microsoft.Storage/storageAccounts/samplestorageaccount/providers/Microsoft.Security/advancedThreatProtectionSettings/current

Design doc for ATP -
https://microsoft.sharepoint.com/teams/Data_Security/Documents/Dev%20Specs/Storage%20Threat%20Detection/Threat%20Detection%20for%20Azure%20Storage%20-%20Design%20Document.docx?d=w996c55b5a160468c9a9bd0434418a4d0

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-java

chlahav commented on specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/security.json:
```diff
@@ -727,6 +727,116 @@
+              }
+            }
+          },
+          "delete": {
```
delete doesn't have a lot of meaning here
what's the difference between this and "isEnabled": false?

please pick one of them (is enabled property or create delete a resource)
I suggest the property since it will allow you to use policies

tomerlmsft replied:
make sense, I removed the DELETE operation

chlahav commented on specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/examples/AdvancedThreatProtection/DeleteAdvancedThreatProtectionSettings_example.json:
```diff
@@ -0,0 +1,11 @@
+    "settingName": "current"
+  },
+  "responses": {
+    "200": {
```
missing 204 example (copy from the JIT delete)

you also have a test that fails on that

tomerlmsft replied:
I'm removing the DELETE operation and stays only with GET and PUT.
When no settings are exists, GET will return "isEnalbed : false" value.
Using PUT, we'll be able to turn the settings ON or OFF.

chlahav commented on specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/security.json:
```diff
@@ -1118,6 +1228,30 @@
+        },
+        "AdvancedThreatProtectionProperties": {
+          "properties": {
+            "isEnabled": {
```
rename to "enabled"

tomerlmsft replied:
ok

chlahav commented on specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/security.json:
```diff
@@ -1235,6 +1369,32 @@
             "description": "name of the Compliance",
             "x-ms-parameter-location": "method"
+        },
+        "ResourceUri": {
```
change to "resourceId" to use the same standard as other RPs

tomerlmsft replied:
thanks

chlahav commented:
please make the "MODE=model PR_ONLY=true" test in Travis pass

marstr commented:
To @chlahav's point, there are some missing and inconsistent examples. You can find details caught by our linter here: https://travis-ci.org/Azure/azure-rest-api-specs/jobs/396930169#L748

marstr commented on specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/security.json:
```diff
@@ -727,6 +727,116 @@
+              "200": {
+                "description": "Successful request to remove Advanced Threat Protection setting."
+              },
+              "204": {
```
Just out of curiosity, is there any scenario where this operation would return a 404?

For example:
200 -> Deleted a resource, it's former contents are returned in this response
204 -> This resource was just deleted by another request.
404 -> There was never a resource by this name.

I'm just not sure what the right REST idiom is here.

tomerlmsft replied:
I removed the DELETE operation and stayed only with PUT (following Chen feedback, see above)

tomerlmsft commented on specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/security.json:
```diff
@@ -727,6 +727,116 @@
+              "200": {
+                "description": "Successful request to get Advanced Threat Protection settings.",
+                "schema": {
+                  "$ref": "#/definitions/AdvancedThreatProtectionSettings"
```
Regarding AdvancedThreatProtection vs AdvancedThreatProtectionSettings - 
We reviewed it internally and I changed to AdvancedThreatProtectionSettings  following a feedback I got.
The purpose of these type is indeed settings so it make sense to me more than AdvancedThreatProtection which doesn't really imply on the real intent.
Also here, all previous features we released followed this pattern - 
VulnerabilityAssessmentSettings, auditingSettings.
and in ASC - workspaceSettings, autoProvisioningSettings, etc...

chlahav replied:
I talked about the object name and not the resource type
the object that you return is only one setting and not settings

in workspaceSettings etc. it's the same

tomerlmsft commented on specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/security.json:
```diff
@@ -727,6 +727,116 @@
+            },
+            "tags": [ "AdvancedThreatProtection" ],
+            "description": "Creates or updates the Advanced Threat Protection settings on a specified resource.",
+            "operationId": "AdvancedThreatProtection_Put",
```
ok

tomerlmsft committed:
Addressing pull request comments

*Removed DELETE operation, stayed only with GET and PUT.
*Changed ResourceUri to ResourceId
*Changed isEnabled to enabled

tomerlmsft committed:
Renaming obejct from AdvancedThreatProtectionSettings to AdvancedThreatProtectionSetting

tomerlmsft committed:
Removed delete operation from the swagger definitions

tomerlmsft committed:
Fixed test failures

marstr commented:
@tomerlmsft, this looks good from my perspective. Given that we've waited for more than a three business days without the ARM team responding, we can go ahead and merge it. However, doing so means you'll potentially have ARM feedback be reported on a production service. This can cause complications. However, it is up to you. Would you like to proceed and have me merge this PR?

ravbhatnagar commented on specification/security/resource-manager/Microsoft.Security/preview/2017-08-01-preview/security.json:
```diff
@@ -1235,6 +1334,32 @@
+          "description": "The identifier of the resource.",
+          "x-ms-parameter-location": "method"
+        },
+        "AdvancedThreatProtectionSettingName": {
```
if current is the only value supported for the setting name we should make an enum with one value.

ravbhatnagar replied:
Also do you support a list API?

tomerlmsft replied:
I changed it to enum with one value, thanks for the comment.

tomerlmsft replied:
We don't support LIST in our API, the only available type of settings name right now is "current".

tomerlmsft committed:
Changing "AdvancedThreatProtectionSettingName" parameter to enum type

tomerlmsft commented:
Hi Mark,
We are still waiting with the merge until the arm manifest will be updated.

marstr commented:
Added the [DoMergeLabel] to indicate, without opening the PR, that we are waiting for the ARM manifest to be updated.

marstr commented:
I'm going to be on vacation for the next two days (until Monday, July 30th). If this PR needs attention in the mean time, please contact my boss, @salameer.

azuresdkci commented:
Can one of the admins verify this patch?

marstr commented:
Any updates on this, @tomerlmsft?

marstr commented:
Closing as stale.

marstr commented:
After an offline request by @tomerlmsft, I'm going to reopen this.

tomerlmsft committed:
Changed param name from 'enabled' to 'isEnabled'

marstr commented:
Restarted CI because it looks like the failure on the most recent commit was unrelated to your change.

marstr commented:
Just as a heads up, I'm going to be OOF all next week. However, since this PR is just waiting for the service to be available, I don't think there's any need to reassign while I'm gone. If you need a swagger reviewer to take a look at this, try getting a hold of @salameer or the swagger reviewer email alias.

marstr commented:
@tomerlmsft, heads up on this conflict that has emerged.

tomerlmsft committed:
Merge branch 'master' into master
